README.md
# Mbox mail analysis
![test status](https://github.com/jacopofar/mailbox-analysis/actions/workflows/test.yaml/badge.svg)
![test status](https://github.com/jacopofar/mailbox-analysis/actions/workflows/lint.yaml/badge.svg)

This is a script that analyzes a **mbox** mail export, such as the one provided by Google Takeout from a Gmail box, and produces a report on the content.

![an heatmap representation of number of mail per day and hour](https://raw.githubusercontent.com/jacopofar/mailbox-report-generator/main/heatmap.png)
![an interactive chart showing the mail activity per day](https://raw.githubusercontent.com/jacopofar/mailbox-report-generator/main/timeline.png)

## Current reports

* received mails over hour of the day and day of the week
* mail per day over time
* most active addresses

## Usage

You need an export of your mailbox in *mbox* format (for Gmail you can get it from [Google Takeout](https://takeout.google.com/)).

Install the tool using pip:

    python3 -m pip install mailbox-report-generator

Then run this command:

    generate_mbox_report "/path/to/the/mbox/file.mbox"

a report will be created in the form of an HTML file, and opened with your default browser.

## Extending the report

The report is generated by running every message through a series of `Processor`s.
Each Processor implements its own logic to aggregate relevant details and can output the report as an HTML string. They are independent from each other.

These strings are simply concatenated to generate a static HTML file, two processors output the header and footer of this file.

This structure makes it quite easy to add or remove specific analysis, run automated tests and implement caching.

# Possible future improvements

- [ ] Examine the mail lenght and word usage over time
  - Note that extracting text from mails is very hard, the multipart format and the weird formats used by advertisement e-mails make it an extremely unreliable operation.
- [ ] Examine the textual content of the emails with SpaCy, retrieve Named Entities like people and locations (see note on previous point)
- [ ] Find which languages are used in the e-mail body

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.1.3] - 2023-02-01

### Changed

- Images in the PyPi page and typos, nothing different in the code

## [0.1.2] - 2023-02-01

### Changed

- Allow user to specify how many addresses to keep in the "most frequent" list
- Aggregate less frequent addresses under "others" to keep the chart readable
## [0.1.1] - 2023-01-26

### Changed

- Display hours in AM/PM format
- Improve display at different screen sizes
- Add README to package description

## [0.1.0] - 2023-01-26

### Changed

- Published as a package to Pypi



tests/unit/test_hour_heatmap.py

from mailbox import Message

from mailanalysis.processors.dow_hour_heatmap import DowHourHeatmap


def test_happy_path():
    p = DowHourHeatmap()
    msg = Message()
    msg.set_payload('Hello world!', 'utf-8')
    msg["Date"] = 'Fri, 9 Jun 2006 12:34:56 -0200'
    p.process(msg)
    generated = p.report_snippet()
    assert '"type":"heatmap"' in generated
    assert '"Mon","Tue","Wed"' in generated
    hrs = ['0'] * 24
    assert generated.count(','.join(hrs)) == 6
    # element 12 + 2 is 1, because it is UTC
    hrs[14] = '1'
    assert str(','.join(hrs)) in generated


def test_empty_mbox():
    p = DowHourHeatmap()
    generated = p.report_snippet()
    assert '"type":"heatmap"' in generated
    assert '"Mon","Tue","Wed"' in generated
    # just ensure it's empty all 7 days of the week
    hrs = ['0'] * 24
    assert generated.count(','.join(hrs)) == 7



tests/unit/test_activity_over_time.py

from mailbox import Message

from mailanalysis.processors.activity_over_time import ActivityOverTime


def test_happy_path():
    p = ActivityOverTime()
    msg = Message()
    msg.set_payload('Hello world!', 'utf-8')
    msg["Date"] = 'Fri, 9 Jun 2006 12:34:56 -0200'
    p.process(msg)
    generated = p.report_snippet()
    assert '"x":["2006-06-09"],"y":[1]' in generated
    msg = Message()
    # no payload this time
    msg["Date"] = 'Sun, 13 Jun 2006 12:34:56 -0200'
    p.process(msg)
    generated = p.report_snippet()
    assert '"x":["2006-06-09","2006-06-13"],"y":[1,1]' in generated


def test_empty_mbox():
    p = ActivityOverTime()
    generated = p.report_snippet()
    assert '"type":"scatter"' in generated
    assert '"x":[]' in generated
    assert '"y":[]' in generated


mailanalysis/__main__.py
import argparse
import importlib.resources as pkg_resources
import logging
import webbrowser
from pathlib import Path

import mailanalysis
import mailanalysis.static_assets
from mailanalysis.processors import ReportHeader
from mailanalysis.processors.activity_over_time import ActivityOverTime
from mailanalysis.processors.dow_hour_heatmap import DowHourHeatmap
from mailanalysis.processors.most_frequent_addresses import MostFrequentAddresses

logger = logging.getLogger(__name__)
logging.basicConfig(
    level=logging.DEBUG,
    format="%(asctime)s %(levelname)s %(name)s %(message)s",
)


def main() -> None:
    parser = argparse.ArgumentParser(
        description="Generate a report on a Mailbox file content."
    )
    parser.add_argument(
        "mbox_file",
        type=str,
        help="The mailbox (.mbox) file to analyze",
    )
    parser.add_argument(
        "--report_file",
        type=str,
        default="report_mail.html",
        help="The report to produce",
    )
    args = parser.parse_args()
    processors = [
        ReportHeader(args.mbox_file),
        DowHourHeatmap(),
        ActivityOverTime(),
        MostFrequentAddresses(most_common_n=25),
    ]
    report_content = mailanalysis.process_mbox(
        args.mbox_file,
        processors,
    )
    header_text = pkg_resources.read_text(
        mailanalysis.static_assets, "html_header.html"
    )
    footer_text = pkg_resources.read_text(
        mailanalysis.static_assets, "html_header.html"
    )
    with open(args.report_file, "w") as f:
        f.write(header_text)
        f.write(report_content)
        f.write(footer_text)
    target = f"file://{Path(args.report_file).absolute()}"
    logger.info(f"Report saved at {target}")
    webbrowser.open(target)


if __name__ == "__main__":
    main()


mailanalysis/processors/__init__.py
import re
from datetime import datetime, timezone
from email.utils import parsedate_to_datetime
from mailbox import Message

# sometimes the mbox files contain weird date formats
# here some cases either ignored or worked around

# 05-08-2005
DD_MM_YYYY = re.compile(r"\d{2}-\d{2}-\d{4}")
# 2016-08-03 18:21:32.174623604 +0000 UTC
ALMOST_ISO = re.compile(r"\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}(\.\d+)? \+0+ UTC")


def parse_mail_date(datestr: str) -> datetime:
    """Parse a date string as a datetime object.

    This handles some quirks found in real mbox files, the RFC regarding the
    date format is not always followed it seems.

    """
    if DD_MM_YYYY.match(datestr):
        raise TypeError(f"Missing time in string {datestr}")
    if ALMOST_ISO.match(datestr):
        return datetime.strptime(datestr[:19], "%Y-%m-%d %H:%M:%S").astimezone(
            timezone.utc
        )
    return parsedate_to_datetime(datestr).astimezone(timezone.utc)


class Processor:
    """Base class for a mail report processor.

    This report generator is designed as a pipeline of processors,
    each processor is created and observes all the messages, then generates
    its own piece of HTML.

    All these snippets are put together with some essential boilerplate to
    form a complete report.
    """

    def process(self, msg: Message) -> None:
        """Process a message, integrating it in the current state.

        Parameters
        ----------
        msg : Message
            A message from a mailbox
        """
        raise NotImplementedError("process method not implemented")

    def report_snippet(self) -> str:
        """Generate the HTML snippet for the specific processor.

        Returns
        -------
        str
            An HTML snippet representing the result of the aggregation
        """
        raise NotImplementedError("report_snippet method not implemented")


class ReportHeader(Processor):
    """Processor for the report introduction text."""

    def __init__(self, filename: str) -> None:
        self.filename = filename
        self.count = 0

    def process(self, msg: Message) -> None:
        self.count += 1

    def report_snippet(self) -> str:
        return f"""
        <section style="padding: 1em;font-family: sans-serif;">
            <h3>Report from {self.filename}</h3>
            <p>Report generated on {datetime.now().date().isoformat()}</p>
            <p>Messages present in the file: {self.count:,.0f}.
            This includes sent messages and spam.<p>
        <section>
        """


mailanalysis/processors/dow_hour_heatmap.py
from mailbox import Message

import plotly.graph_objects as go

from mailanalysis.processors import Processor, parse_mail_date


class DowHourHeatmap(Processor):
    """Processor to create a heatmap for day of the week and hour.

    The resulting report snippet shows how many messages there were in a
    specific day of the week and hour, e.g. Tuesday at 15

    The hour is in UTC.
    """

    def __init__(self) -> None:
        self.dow_hour = [[0 for _ in range(24)] for _ in range(7)]

    def process(self, msg: Message) -> None:
        if msg["Date"] is None:
            return
        try:
            msg_date = parse_mail_date(msg["Date"])
        except TypeError:
            print(f'cannot parse date {msg["Date"]}')
            return
        utc_hour = msg_date.time().hour
        dow = msg_date.weekday()
        self.dow_hour[dow][utc_hour] += 1

    def report_snippet(self) -> str:
        dow_hour = self.dow_hour
        fig = go.Figure(
            data=go.Heatmap(
                # trick to hide the extra double label
                hoverlabel=dict(namelength=0),
                hovertemplate="%{y} at %{x}: %{z} messages",
                x=(
                    ["12 AM"]
                    + [f"{h} AM" for h in range(1, 12)]
                    + ["12 PM"]
                    + [f"{h} PM" for h in range(1, 12)]
                ),
                y=["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"],
                z=dow_hour,
            ),
        )
        fig.update_layout(title="Messages by UTC hour and day of the week")
        # reading from top to bottom the days of the weekß
        fig.update_yaxes(autorange="reversed")

        return fig.to_html(full_html=False, include_plotlyjs=False)  # type: ignore


mailanalysis/processors/activity_over_time.py
from datetime import date
from mailbox import Message

import plotly.graph_objects as go

from mailanalysis.processors import Processor, parse_mail_date


class ActivityOverTime(Processor):
    """Processor to create an histogram of activity over time.

    It shows how many mails were exchanged per day over time.
    """

    def __init__(self) -> None:
        self.mail_per_day: dict[date, int] = {}

    def process(self, msg: Message) -> None:
        if msg["Date"] is None:
            return
        try:
            msg_date = parse_mail_date(msg["Date"]).date()
        except TypeError:
            print(f'cannot parse date {msg["Date"]}')
            return
        self.mail_per_day[msg_date] = self.mail_per_day.get(msg_date, 0) + 1

    def report_snippet(self) -> str:
        dates = sorted(self.mail_per_day.keys())
        values = [self.mail_per_day[d] for d in dates]
        fig = go.Figure(
            data=go.Scatter(
                x=dates,
                y=values,
            )
        )
        fig.update_layout(title="Messages by UTC hour and day of the week")
        # reading from top to bottom the days of the weekß
        fig.update_layout(
            title_text="Mail per day",
            xaxis_rangeslider_visible=True,
        )

        return fig.to_html(full_html=False, include_plotlyjs=False)  # type: ignore


mailanalysis/processors/most_frequent_addresses.py
from collections import Counter
from email.utils import parseaddr
from mailbox import Message

import plotly.graph_objects as go

from mailanalysis.processors import Processor


class MostFrequentAddresses(Processor):
    """Processor for the most frequent addresses.

    To keep the chart readable only the 20 more frequent are shown
    """

    def __init__(self, most_common_n: int = 20) -> None:
        self.address_count: Counter[str] = Counter()
        self.most_common_n = most_common_n

    def process(self, msg: Message) -> None:
        if msg["From"] is None or msg["To"] is None:
            return
        # if the string is weird we get an Header object, so by forcing
        # the conversion to str it goes back to a parsable mail address
        from_addr = parseaddr(str(msg["From"]))[1].lower()
        self.address_count.update([from_addr])
        to_addr = parseaddr(str(msg["To"]))[1].lower()
        self.address_count.update([to_addr])

    def report_snippet(self) -> str:
        # ignore address #1 assuming it is the mailbox owner
        top_addresses = self.address_count.most_common(self.most_common_n)[1:]
        top_addresses.append(
            (
                "(others)",
                sum(
                    v for _, v in self.address_count.most_common()[self.most_common_n :]
                ),
            )
        )
        labels, values = zip(*top_addresses)
        fig = go.Figure(
            data=go.Pie(
                values=values,
                labels=labels,
            )
        )
        # reading from top to bottom the days of the weekß
        fig.update_layout(
            title_text="Most active addresses",
        )

        return fig.to_html(full_html=False, include_plotlyjs=False)  # type: ignore


mailanalysis/processors.py
import re
from collections import Counter
from datetime import date, datetime, timezone
from email.utils import parseaddr, parsedate_to_datetime
from mailbox import Message

import plotly.graph_objects as go

# sometimes the mbox files contain weird date formats
# here some cases either ignored or worked around

# 05-08-2005
DD_MM_YYYY = re.compile(r"\d{2}-\d{2}-\d{4}")
# 2016-08-03 18:21:32.174623604 +0000 UTC
ALMOST_ISO = re.compile(r"\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}(\.\d+)? \+0+ UTC")


def parse_mail_date(datestr: str) -> datetime:
    """Parse a date string as a datetime object.

    This handles some quirks found in real mbox files, the RFC regarding the
    date format is not always followed it seems.

    """
    if DD_MM_YYYY.match(datestr):
        raise TypeError(f"Missing time in string {datestr}")
    if ALMOST_ISO.match(datestr):
        return datetime.strptime(datestr[:19], "%Y-%m-%d %H:%M:%S").astimezone(
            timezone.utc
        )
    return parsedate_to_datetime(datestr).astimezone(timezone.utc)


class Processor:
    """Base class for a mail report processor.

    This report generator is designed as a pipeline of processors,
    each processor is created and observes all the messages, then generates
    its own piece of HTML.

    All these snippets are put together with some essential boilerplate to
    form a complete report.
    """

    def process(self, msg: Message) -> None:
        """Process a message, integrating it in the current state.

        Parameters
        ----------
        msg : Message
            A message from a mailbox
        """
        raise NotImplementedError("process method not implemented")

    def report_snippet(self) -> str:
        """Generate the HTML snippet for the specific processor.

        Returns
        -------
        str
            An HTML snippet representing the result of the aggregation
        """
        raise NotImplementedError("report_snippet method not implemented")


class DowHourHeatmap(Processor):
    """Processor to create a heatmap for day of the week and hour.

    The resulting report snippet shows how many messages there were in a
    specific day of the week and hour, e.g. Tuesday at 15

    The hour is in UTC.
    """

    def __init__(self) -> None:
        self.dow_hour = [[0 for _ in range(24)] for _ in range(7)]

    def process(self, msg: Message) -> None:
        if msg["Date"] is None:
            return
        try:
            msg_date = parse_mail_date(msg["Date"])
        except TypeError:
            print(f'cannot parse date {msg["Date"]}')
            return
        utc_hour = msg_date.time().hour
        dow = msg_date.weekday()
        self.dow_hour[dow][utc_hour] += 1

    def report_snippet(self) -> str:
        dow_hour = self.dow_hour
        fig = go.Figure(
            data=go.Heatmap(
                # trick to hide the extra double label
                hoverlabel=dict(namelength=0),
                hovertemplate="%{y} at %{x}: %{z} messages",
                x=[h for h in range(24)],
                y=["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"],
                z=dow_hour,
            ),
        )
        fig.update_layout(title="Messages by UTC hour and day of the week")
        # reading from top to bottom the days of the weekß
        fig.update_yaxes(autorange="reversed")

        return fig.to_html(full_html=False, include_plotlyjs=False)  # type: ignore


class ActivityOverTime(Processor):
    """Processor to create an histogram of activity over time.

    It shows how many mails were exchanged per day over time.
    """

    def __init__(self) -> None:
        self.mail_per_day: dict[date, int] = {}

    def process(self, msg: Message) -> None:
        if msg["Date"] is None:
            return
        try:
            msg_date = parse_mail_date(msg["Date"]).date()
        except TypeError:
            print(f'cannot parse date {msg["Date"]}')
            return
        self.mail_per_day[msg_date] = self.mail_per_day.get(msg_date, 0) + 1

    def report_snippet(self) -> str:
        dates = sorted(self.mail_per_day.keys())
        values = [self.mail_per_day[d] for d in dates]
        fig = go.Figure(
            data=go.Scatter(
                x=dates,
                y=values,
            )
        )
        fig.update_layout(title="Messages by UTC hour and day of the week")
        # reading from top to bottom the days of the weekß
        fig.update_layout(
            title_text="Mail per day",
            xaxis_rangeslider_visible=True,
        )

        return fig.to_html(full_html=False, include_plotlyjs=False)  # type: ignore


class MostFrequentAddresses(Processor):
    """Processor for the most frequent addresses."""

    def __init__(self) -> None:
        self.address_count: Counter[str] = Counter()

    def process(self, msg: Message) -> None:
        if msg["From"] is None or msg["To"] is None:
            return
        # if the string is weird we get an Heade ronject, so by forcing
        # the conversion to str it goes back to a parsable mail address
        from_addr = parseaddr(str(msg["From"]))[1].lower()
        self.address_count.update([from_addr])
        to_addr = parseaddr(str(msg["To"]))[1].lower()
        self.address_count.update([to_addr])

    def report_snippet(self) -> str:
        # ignore address #1 assuming it is the mailbox owner
        top_addresses = self.address_count.most_common(20)[1:]
        labels, values = zip(*top_addresses)
        fig = go.Figure(
            data=go.Pie(
                values=values,
                labels=labels,
            )
        )
        # reading from top to bottom the days of the weekß
        fig.update_layout(
            title_text="Most active addresses",
        )

        return fig.to_html(full_html=False, include_plotlyjs=False)  # type: ignore


class ReportHeader(Processor):
    """Processor for the report introduction text."""

    def __init__(self, filename: str) -> None:
        self.filename = filename
        self.count = 0

    def process(self, msg: Message) -> None:
        self.count += 1

    def report_snippet(self) -> str:
        return f"""
        <section style="padding: 1em;font-family: sans-serif;">
            <h3>Report from {self.filename}</h3>
            <p>Report generated on {datetime.now().date().isoformat()}</p>
            <p>Messages present in the file: {self.count:,.0f}.
            This includes sent messages and spam.<p>
        <section>
        """
